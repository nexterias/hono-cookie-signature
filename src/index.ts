export {
  getVerifiedCookie,
  setCookieWithSignature,
  getCookie,
} from './functions'
export { cookieSignature } from './middleware'
export { importKey } from './signature'
