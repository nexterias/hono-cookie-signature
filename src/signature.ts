export const HASH_NAME = 'HMAC'
export const HASH_ALGORITHM = 'SHA-256'

/**
 * Imports a key from a buffer source.
 *
 * @param key The buffer source containing the key data.
 * @returns A promise that resolves with the imported key.
 *
 * @example
 * ```ts
 * import { importKey } from '@nexterias/hono-cookie-signature'
 *
 * const secret = await importKey(new TextEncoder().encode('THIS_IS_ULTRA_HYPER_SECRET_KEY'))
 *
 * console.log(secret)
 * ```
 */
export const importKey = (key: BufferSource) => {
  return crypto.subtle.importKey(
    'raw',
    key,
    { name: HASH_NAME, hash: HASH_ALGORITHM },
    false,
    ['sign', 'verify']
  )
}

/**
 * Signs the given data with the provided key.
 *
 * @param data The data to sign.
 * @param key The key to use for signing.
 * @returns A promise that resolves with the signature as a Uint8Array.
 */
export const sign = async (data: BufferSource, key: CryptoKey) => {
  const buf = await crypto.subtle.sign(
    { name: HASH_NAME, hash: HASH_ALGORITHM },
    key,
    data
  )

  return new Uint8Array(buf) // ArrayBuffer to Uint8Array
}

/**
 * Verifies the given signature for the provided data and key.
 *
 * @param key The key to use for verification.
 * @param signature The signature to verify.
 * @param data The source data to verify.
 * @returns A promise that resolves with a boolean indicating whether the signature is valid.
 */
export const verify = (
  key: CryptoKey,
  signature: BufferSource,
  data: BufferSource
) =>
  crypto.subtle.verify(
    { name: HASH_NAME, hash: HASH_ALGORITHM },
    key,
    signature,
    data
  )
