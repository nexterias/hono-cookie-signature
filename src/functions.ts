import { Context as HonoContext } from 'hono'
import { setCookie } from 'hono/cookie'
import type { CookieOptions } from 'hono/utils/cookie'
import { sign } from './signature'
import { encodeBase64Url } from './utils/base64url'
import { getContextKey } from './utils/context'
import { concatCookieValue } from './utils/cookie'

export const setCookieWithSignature =
  (key: CryptoKey) =>
  async (
    context: HonoContext,
    name: string,
    value: string,
    options?: CookieOptions
  ) => {
    const encoder = new TextEncoder()
    const valueBuf = encoder.encode(value)
    const signature = await sign(valueBuf, key)

    setCookie(
      context,
      name,
      concatCookieValue(encodeBase64Url(valueBuf), encodeBase64Url(signature)),
      options
    )

    return signature
  }

export const getVerifiedCookie = (context: HonoContext, name: string) => {
  const value = context.get(getContextKey(name))

  if (typeof value !== 'string') return
  else return value
}
