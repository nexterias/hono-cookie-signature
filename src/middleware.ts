import type { MiddlewareHandler } from 'hono'
import { getCookie } from 'hono/cookie'
import { decodeBase64Url } from './utils/base64url'
import { verify } from './signature'
import { getContextKey } from './utils/context'
import { splitCookieValue } from './utils/cookie'

/**
 * Options for cookie signature middleware.
 */
export interface CookieSignatureOptions {
  /**
   * The secret key used to sign the cookie.
   */
  secret: CryptoKey

  /**
   * An array of cookie keys to verify the signature for.
   */
  cookies: string[]
}

/**
 * @param options The options for cookie signature middleware.
 * @returns Middleware handler.
 */
export const cookieSignature = (
  options: Readonly<CookieSignatureOptions>
): MiddlewareHandler => {
  return async (context, next) => {
    for (const cookieKey of options.cookies) {
      const rawCookieValue = getCookie(context, cookieKey)
      if (!rawCookieValue)
        return context.json(
          { message: `Cookie(${cookieKey}) is required.` },
          400
        )

      const splittedCookieValue = splitCookieValue(rawCookieValue)
      if (!splittedCookieValue.value || !splittedCookieValue.signature)
        return context.json({ message: 'Invalid cookie value' }, 400)

      const signature = decodeBase64Url(splittedCookieValue.signature)
      const cookieValue = decodeBase64Url(splittedCookieValue.value)

      const verified = await verify(options.secret, signature, cookieValue)
      if (!verified) return context.json({ message: 'Invalid signature' }, 400)

      const decoder = new TextDecoder()
      context.set(getContextKey(cookieKey), decoder.decode(cookieValue))
    }

    return next()
  }
}
