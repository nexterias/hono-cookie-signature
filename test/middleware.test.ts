import { Hono } from 'hono'
import { expect, test } from 'vitest'
import { cookieSignature } from '../src/middleware'
import { getVerifiedCookie } from '../src/functions'

test('cookieSignature', async () => {
  const secret = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode('THIS_IS_SECRET_KEY'),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign', 'verify']
  )
  const app = new Hono()

  app.get(
    '/cookie',
    cookieSignature({
      cookies: ['bassfreq', 'caz'],
      secret,
    }),
    async (context) => {
      expect(getVerifiedCookie(context, 'bassfreq')).toBe('Rawstyle')
      expect(getVerifiedCookie(context, 'caz')).toBe('MEGATON KICK')
      expect(getVerifiedCookie(context, 'mahiro')).toBeUndefined()

      return context.text('ok')
    }
  )

  {
    // cookiesで設定されたキーのペアが存在しない場合はHTTP 400を返すこと
    const response = await app.request('/cookie')

    expect(response.status).toBe(400)
  }

  {
    // cookiesで設定されたキーがCookieに存在するものの、フォーマットが不正な場合はHTTP 400を返すこと
    const response = await app.request('/cookie', {
      headers: {
        cookie: 'bassfreq=rawstyle; caz=BCM',
      },
    })

    expect(response.status).toBe(400)
  }

  {
    const response = await app.request('/cookie', {
      headers: {
        cookie: [
          'bassfreq=UmF3c3R5bGU.W57skKFxTQuP6er86djSk2cL9iP3h9W-FzePE6auus4',
          'caz=TUVHQVRPTiBLSUNL.k203sS2F7OLETAI9bQgTm8Md0J1yw_VMndVwzjrAxfE',
        ].join('; '),
      },
    })

    expect(response.status).toBe(401)
  }

  {
    // Cookieに設定されている値と署名が正しい場合はHTTP 200を返す
    const response = await app.request('/cookie', {
      headers: {
        cookie: [
          'bassfreq=UmF3c3R5bGU.W57skKFxTQuP6er86djSk2cL9iP3h9W-FzePE6auus4',
          'caz=TUVHQVRPTiBLSUNL.zAcaV3FbkrXArqSUjuAAJJIoyvZxLXP6N7EpG0wkL1g',
        ].join('; '),
      },
    })

    expect(response.status).toBe(200)
  }
})
