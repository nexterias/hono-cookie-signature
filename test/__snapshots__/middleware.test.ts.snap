// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cookieSignature 1`] = `
{
  "message": "Cookie(bassfreq) is required.",
}
`;

exports[`cookieSignature 2`] = `
{
  "message": "Invalid cookie value",
}
`;

exports[`cookieSignature 3`] = `
{
  "message": "Invalid signature",
}
`;
